### PR TITLE
Add ccache / don't run action twice on pull_requests

### PR DESCRIPTION
**Pull request by mgehre-amd**
Brings the CI time to 10 minutes.

Only build on pushes to the main/feature branch. Otherwise every PR sees two builds because pushing to the PR branch and changing the pull_request both trigger it separate builds.

Introduce ccache to speed-up onnx-mlir builds.

### CODE REVIEW

File: .github/workflows/macos-amd64-build.yml [hunk 1]
```diff
@@ -56,7 +66,8 @@ jobs:
       run: |
         cd ~/work/onnx-mlir
         sh ~/work/onnx-mlir/onnx-mlir/utils/check-unittest.sh
-    - name: run onnx-mlir backend and numerical tests
-      run: |
-        cd ~/work/onnx-mlir
-        sh ~/work/onnx-mlir/onnx-mlir/utils/check-onnx-backend-numerical.sh
+    # Those tests are not relevant to the work on the xilinx fork, but take 40 min
```
Comment (TinaAMD) [inline .github/workflows/macos-amd64-build.yml:69]: When would they become relevant? When we would start to modify what the onnx mlir ops look like?

File: .github/workflows/macos-amd64-build.yml [hunk 2]
```diff
@@ -1,6 +1,9 @@
 name: GitHub Action MacOS amd64
 
-on: [push, pull_request]
+on:
```
Comment (TinaAMD) [inline .github/workflows/macos-amd64-build.yml:3]: I guess we could have the same for the format checkers, they seem to be run twice as well?

**Review by TinaAMD — APPROVED** [review]
LGTM

File: .github/workflows/macos-amd64-build.yml (hunk 2 above)
Comment (mgehre-amd) [reply .github/workflows/macos-amd64-build.yml:3]: yes, but at least they are fast

File: .github/workflows/macos-amd64-build.yml (hunk 1 above)
Comment (mgehre-amd) [reply .github/workflows/macos-amd64-build.yml:69]: I think this mostly tests the onnx-mlir -> krnl -> cpu backend path, and not the onnx-mlir -> TOSA path.

Comment (mgehre-amd) [reply .github/workflows/macos-amd64-build.yml:69]: And we cannot modify onnx-mlir ops as they are generated from the onnx spec

Comment (TinaAMD) [reply .github/workflows/macos-amd64-build.yml:69]: Ok, we currently do not work on this, but I was wondering how we would notice if we would start doing so. I guess that could lead to people breaking tests without noticing.

**Review by ttjost — APPROVED** [review]

**Pull request merged**